File: hotel-booking-backend/services/momoService.js
/**
 * MoMo Service - Xử lý tích hợp thanh toán MoMo
 * 
 * Chức năng:
 * - Tạo payment request đến MoMo
 * - Verify signature từ MoMo
 * - Query transaction status
 * - Refund transaction
 */

const crypto = require('crypto');
const https = require('https');
const momoConfig = require('../config/momo');

class MoMoService {
  /**
   * Tạo payment URL từ MoMo (giống VNPay - return object với paymentUrl, qrCodeUrl, deeplink)
   * 
   * @param {Object} params - Thông tin thanh toán
   * @param {string} params.orderId - Mã đơn hàng
   * @param {number} params.amount - Số tiền (VND)
   * @param {string} params.orderInfo - Thông tin đơn hàng
   * @returns {Promise<Object>} Object với {paymentUrl, qrCodeUrl, deeplink}
   */
  async createPaymentUrl(params) {
    const { orderId, amount, orderInfo } = params;
    
    // Gọi createPayment để lấy payUrl và qrCodeUrl
    const result = await this.createPayment({
      orderId,
      amount,
      orderInfo,
      extraData: '',
    });
    
    // Return object với paymentUrl, qrCodeUrl và deeplink
    if (result.payUrl) {
      return {
        paymentUrl: result.payUrl,
        qrCodeUrl: result.qrCodeUrl || null,
        deeplink: result.deeplink || null,
      };
    } else {
      throw new Error('MoMo did not return payment URL');
    }
  }

  /**
   * Tạo payment request đến MoMo
   * 
   * @param {Object} params - Thông tin thanh toán
   * @param {string} params.orderId - Mã đơn hàng
   * @param {number} params.amount - Số tiền (VND)
   * @param {string} params.orderInfo - Thông tin đơn hàng
   * @param {string} params.extraData - Dữ liệu bổ sung (optional, base64 encoded)
   * @returns {Promise<Object>} Response từ MoMo với payUrl
   */
  async createPayment(params) {
    const {
      orderId,
      amount,
      orderInfo,
      extraData = '',
    } = params;

    // Validate required params
    if (!orderId || !amount || !orderInfo) {
      throw new Error('Missing required parameters: orderId, amount, orderInfo');
    }

    // Generate requestId (unique)
    const requestId = momoConfig.partnerCode + new Date().getTime();

    // Tạo raw signature theo format của MoMo API v2
    // QUAN TRỌNG: Thứ tự các field phải SẮP XẾP ALPHABETICALLY theo MoMo API documentation
    // Thứ tự đúng (alphabetically): accessKey, amount, extraData, ipnUrl, orderId, orderInfo, partnerCode, redirectUrl, requestId, requestType
    // Lưu ý: Dùng redirectUrl (không phải returnUrl) và ipnUrl (không phải notifyUrl) trong signature
    const rawSignature = 
      'accessKey=' + momoConfig.accessKey +
      '&amount=' + amount +
      '&extraData=' + (extraData || '') +
      '&ipnUrl=' + momoConfig.ipnUrl +
      '&orderId=' + orderId +
      '&orderInfo=' + orderInfo +
      '&partnerCode=' + momoConfig.partnerCode +
      '&redirectUrl=' + momoConfig.returnUrl +
      '&requestId=' + requestId +
      '&requestType=' + momoConfig.requestType;

    console.log('--------------------RAW SIGNATURE----------------');
    console.log(rawSignature);

    // Tạo signature bằng HMAC SHA256
    const signature = crypto
      .createHmac('sha256', momoConfig.secretKey)
      .update(rawSignature)
      .digest('hex');

    console.log('--------------------SIGNATURE----------------');
    console.log(signature);

    // Request body gửi đến MoMo - THEO ĐÚNG FORMAT MOMO API v2
    // Lưu ý: requestType và lang không có trong signature string, chỉ có trong request body
    const requestBody = JSON.stringify({
      partnerCode: momoConfig.partnerCode,
      accessKey: momoConfig.accessKey,
      requestId: requestId,
      amount: amount.toString(),
      orderId: orderId,
      orderInfo: orderInfo,
      redirectUrl: momoConfig.returnUrl,
      ipnUrl: momoConfig.ipnUrl,
      extraData: extraData,
      requestType: momoConfig.requestType,
      signature: signature,
      lang: momoConfig.lang,
    });

    console.log('--------------------REQUEST BODY----------------');
    console.log(requestBody);

    // Gọi MoMo API với retry logic
    const maxRetries = 2;
    let lastError = null;
    
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        console.log(`📤 Sending request to MoMo API (attempt ${attempt}/${maxRetries}):`, momoConfig.apiEndpoint);
        const response = await this._sendRequest(momoConfig.apiEndpoint, requestBody);
        
        console.log('--------------------MOMO RESPONSE----------------');
        console.log(JSON.stringify(response, null, 2));

        // Check response
        if (response.resultCode === 0) {
          // Success - trả về payUrl để redirect user
          console.log('✅ MoMo payment URL created successfully');
          console.log('   Pay URL:', response.payUrl);
          console.log('   Deep Link:', response.deeplink);
          console.log('   QR Code URL:', response.qrCodeUrl);
          
          return {
            success: true,
            payUrl: response.payUrl,
            deeplink: response.deeplink,
            qrCodeUrl: response.qrCodeUrl,
            requestId: requestId,
            orderId: orderId,
          };
        } else {
          // Error từ MoMo
          const errorMessage = response.message || this.getResultMessage(response.resultCode);
          console.error(`❌ MoMo API Error: resultCode=${response.resultCode}, message=${errorMessage}`);
          throw new Error(`MoMo error (code ${response.resultCode}): ${errorMessage}`);
        }
      } catch (error) {
        lastError = error;
        console.error(`❌ Error calling MoMo API (attempt ${attempt}/${maxRetries}):`);
        console.error('   Error Type:', error.constructor.name);
        console.error('   Error Message:', error.message);
        
        // Nếu là timeout, connection error, hoặc HTTP 5xx error và chưa hết retry, thử lại
        const isRetryableError = 
          error.message.includes('timeout') || 
          error.message.includes('ECONNREFUSED') || 
          error.message.includes('ENOTFOUND') ||
          error.message.includes('ECONNRESET') ||
          error.message.includes('socket hang up') ||
          error.message.includes('502') ||
          error.message.includes('503') ||
          error.message.includes('504') ||
          error.message.includes('Bad Gateway') ||
          error.message.includes('Service Unavailable') ||
          error.message.includes('Gateway Timeout');
        
        if (attempt < maxRetries && isRetryableError) {
          const waitTime = attempt * 2000; // 2s, 4s
          console.log(`⏳ Retrying in ${waitTime}ms... (MoMo server may be temporarily unavailable)`);
          await new Promise(resolve => setTimeout(resolve, waitTime));
          continue;
        }
        
        // Nếu không phải lỗi có thể retry, hoặc đã hết retry, throw error
        break;
      }
    }
    
    // Nếu đến đây, tất cả retry đã thất bại
    console.error('❌ All retry attempts failed');
    console.error('   Last Error:', lastError?.message);
    console.error('   Stack:', lastError?.stack);
    
    // Cải thiện error message dựa trên loại lỗi
    if (lastError?.message.includes('timeout')) {
      throw new Error('MoMo API request timeout sau nhiều lần thử. MoMo server có thể đang quá tải hoặc không phản hồi. Vui lòng thử lại sau hoặc sử dụng phương thức thanh toán khác (VNPay).');
    } else if (lastError?.message.includes('ECONNREFUSED') || lastError?.message.includes('ENOTFOUND')) {
      throw new Error('Không thể kết nối đến MoMo API sau nhiều lần thử. Kiểm tra kết nối mạng hoặc API endpoint.');
    } else if (lastError?.message.includes('ECONNRESET') || lastError?.message.includes('socket hang up')) {
      throw new Error('Kết nối đến MoMo API bị ngắt. MoMo server có thể đang quá tải. Vui lòng thử lại sau hoặc sử dụng phương thức thanh toán khác (VNPay).');
    } else if (lastError?.message.includes('502') || lastError?.message.includes('Bad Gateway')) {
      throw new Error('MoMo payment gateway đang tạm thời không khả dụng (502 Bad Gateway). Vui lòng thử lại sau hoặc sử dụng phương thức thanh toán khác.');
    } else if (lastError?.message.includes('HTML')) {
      throw new Error('MoMo API trả về lỗi. Kiểm tra credentials và API endpoint trong file .env.');
    } else {
      throw lastError || new Error('Unknown error calling MoMo API');
    }
  }

  /**
   * Verify signature từ MoMo IPN/Return
   * 
   * @param {Object} data - Data từ MoMo callback
   * @returns {boolean} True nếu signature hợp lệ
   */
  verifySignature(data) {
    const {
      partnerCode,
      orderId,
      requestId,
      amount,
      orderInfo,
      orderType,
      transId,
      resultCode,
      message,
      payType,
      responseTime,
      extraData,
      signature,
    } = data;

    // Tạo raw signature để verify - THEO ĐÚNG THỨ TỰ CỦA MOMO API v2
    // Thứ tự cho verify: partnerCode, accessKey, requestId, amount, orderId, orderInfo, returnUrl, notifyUrl, extraData
    // Nhưng trong response có thêm các field: message, orderType, payType, responseTime, resultCode, transId
    // Cần kiểm tra documentation để biết thứ tự chính xác cho verify
    // Tạm thời dùng thứ tự: partnerCode, accessKey, requestId, amount, orderId, orderInfo, returnUrl, notifyUrl, extraData, message, orderType, payType, responseTime, resultCode, transId
    const rawSignature =
      'partnerCode=' + (partnerCode || momoConfig.partnerCode) +
      '&accessKey=' + momoConfig.accessKey +
      '&requestId=' + (requestId || '') +
      '&amount=' + (amount || '') +
      '&orderId=' + (orderId || '') +
      '&orderInfo=' + (orderInfo || '') +
      '&returnUrl=' + momoConfig.returnUrl +
      '&notifyUrl=' + momoConfig.ipnUrl +
      '&extraData=' + (extraData || '') +
      '&message=' + (message || '') +
      '&orderType=' + (orderType || '') +
      '&payType=' + (payType || '') +
      '&responseTime=' + (responseTime || '') +
      '&resultCode=' + (resultCode || '') +
      '&transId=' + (transId || '');

    console.log('--------------------VERIFY RAW SIGNATURE----------------');
    console.log(rawSignature);

    // Tạo signature để so sánh
    const expectedSignature = crypto
      .createHmac('sha256', momoConfig.secretKey)
      .update(rawSignature)
      .digest('hex');

    console.log('Expected Signature:', expectedSignature);
    console.log('Received Signature:', signature);

    return signature === expectedSignature;
  }

  /**
   * Query transaction status từ MoMo
   * 
   * @param {Object} params - Thông tin query
   * @param {string} params.orderId - Mã đơn hàng
   * @param {string} params.requestId - Request ID từ lúc tạo payment
   * @returns {Promise<Object>} Kết quả query
   */
  async queryTransaction(params) {
    const { orderId, requestId } = params;

    // Tạo raw signature cho query transaction - THEO ĐÚNG THỨ TỰ MOMO API v2
    // Thứ tự: partnerCode, accessKey, requestId, orderId
    const rawSignature =
      'partnerCode=' + momoConfig.partnerCode +
      '&accessKey=' + momoConfig.accessKey +
      '&requestId=' + requestId +
      '&orderId=' + orderId;

    const signature = crypto
      .createHmac('sha256', momoConfig.secretKey)
      .update(rawSignature)
      .digest('hex');

    const requestBody = JSON.stringify({
      partnerCode: momoConfig.partnerCode,
      accessKey: momoConfig.accessKey,
      requestId: requestId,
      orderId: orderId,
      signature: signature,
      lang: momoConfig.lang,
    });

    // MoMo query endpoint
    const queryEndpoint = 'https://test-payment.momo.vn/v2/gateway/api/query';

    try {
      const response = await this._sendRequest(queryEndpoint, requestBody);
      return response;
    } catch (error) {
      console.error('Error querying MoMo transaction:', error);
      throw error;
    }
  }

  /**
   * Gửi HTTPS request đến MoMo
   * 
   * @param {string} url - MoMo API endpoint
   * @param {string} body - Request body (JSON string)
   * @returns {Promise<Object>} Response từ MoMo
   */
  _sendRequest(url, body) {
    return new Promise((resolve, reject) => {
      const urlObj = new URL(url);
      
      const options = {
        hostname: urlObj.hostname,
        port: 443,
        path: urlObj.pathname,
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(body),
          'Connection': 'keep-alive', // Giữ kết nối để tăng tốc độ
        },
        timeout: 60000, // 60 seconds timeout
      };

      const req = https.request(options, (res) => {
        let data = '';

        // Log response status và headers
        console.log('--------------------MOMO API RESPONSE----------------');
        console.log('Status Code:', res.statusCode);
        console.log('Status Message:', res.statusMessage);
        console.log('Headers:', JSON.stringify(res.headers, null, 2));

        // Kiểm tra HTTP status code
        if (res.statusCode < 200 || res.statusCode >= 300) {
          console.error(`❌ MoMo API returned error status: ${res.statusCode} ${res.statusMessage}`);
          
          // Log upstream status nếu có (từ APISIX gateway)
          if (res.headers['x-apisix-upstream-status']) {
            console.error(`   Upstream Status: ${res.headers['x-apisix-upstream-status']}`);
            console.error('   💡 This indicates MoMo backend servers are not responding');
          }
        }

        res.on('data', (chunk) => {
          data += chunk;
        });

        res.on('end', () => {
          console.log('Response Body (raw):');
          console.log(data);
          console.log('Response Body Length:', data.length);
          
          // Kiểm tra nếu response rỗng
          if (!data || data.trim().length === 0) {
            console.error('❌ MoMo returned empty response');
            reject(new Error(`MoMo returned empty response. Status: ${res.statusCode}`));
            return;
          }

          // Kiểm tra nếu response là HTML (thường là error page)
          if (data.trim().startsWith('<!DOCTYPE') || data.trim().startsWith('<html')) {
            console.error('❌ MoMo returned HTML instead of JSON (likely an error page)');
            console.error('HTML Response (first 500 chars):', data.substring(0, 500));
            reject(new Error(`MoMo returned HTML error page. Status: ${res.statusCode}. Check MoMo API endpoint and credentials.`));
            return;
          }

          try {
            const jsonResponse = JSON.parse(data);
            console.log('✅ Successfully parsed JSON response:');
            console.log(JSON.stringify(jsonResponse, null, 2));
            
            // Kiểm tra nếu có error trong JSON response
            if (jsonResponse.resultCode && jsonResponse.resultCode !== 0) {
              const errorMessage = jsonResponse.message || this.getResultMessage(jsonResponse.resultCode);
              console.error(`❌ MoMo API Error: resultCode=${jsonResponse.resultCode}, message=${errorMessage}`);
            }
            
            resolve(jsonResponse);
          } catch (e) {
            console.error('❌ Failed to parse JSON response:');
            console.error('Parse Error:', e.message);
            console.error('Response Data:', data);
            reject(new Error(`Invalid JSON response from MoMo. Status: ${res.statusCode}. Response: ${data.substring(0, 200)}`));
          }
        });
      });

      // Handle request errors (network errors, timeouts, etc.)
      req.on('error', (error) => {
        console.error('❌ Request error:', error);
        reject(error);
      });

      // Set timeout (60 seconds - tăng từ 30s để tránh timeout quá nhanh)
      req.setTimeout(60000, () => {
        console.error('⏱️ Request timeout after 60 seconds');
        req.destroy();
        reject(new Error('MoMo API request timeout after 60 seconds. MoMo server có thể đang quá tải hoặc không phản hồi.'));
      });

      req.on('error', (error) => {
        reject(error);
      });

      req.write(body);
      req.end();
    });
  }

  /**
   * Parse result code từ MoMo
   * 
   * @param {number} resultCode - Result code từ MoMo
   * @returns {string} Thông báo tương ứng
   */
  getResultMessage(resultCode) {
    const messages = {
      0: 'Giao dịch thành công',
      9000: 'Giao dịch được khởi tạo, chờ người dùng xác nhận thanh toán',
      8000: 'Giao dịch đang được xử lý',
      7000: 'Giao dịch đang chờ thanh toán',
      1000: 'Giao dịch đã được khởi tạo, chờ người dùng xác nhận thanh toán',
      11: 'Truy cập bị từ chối',
      12: 'Phiên bản API không được hỗ trợ cho yêu cầu này',
      13: 'Xác thực dữ liệu thất bại',
      20: 'Số tiền không hợp lệ',
      21: 'Số tiền thanh toán không hợp lệ',
      40: 'RequestId bị trùng',
      41: 'OrderId bị trùng',
      42: 'OrderId không hợp lệ hoặc không được tìm thấy',
      43: 'Yêu cầu bị từ chối vì xung đột trong quá trình xử lý giao dịch',
      1001: 'Giao dịch thanh toán thất bại do tài khoản người dùng không đủ tiền',
      1002: 'Giao dịch bị từ chối do nhà phát hành tài khoản thanh toán',
      1003: 'Giao dịch bị hủy',
      1004: 'Giao dịch thất bại do số tiền thanh toán vượt quá hạn mức thanh toán của người dùng',
      1005: 'Giao dịch thất bại do url hoặc QR code đã hết hạn',
      1006: 'Giao dịch thất bại do người dùng đã từ chối xác nhận thanh toán',
      1007: 'Giao dịch bị từ chối vì tài khoản người dùng đang ở trạng thái tạm khóa',
      1026: 'Giao dịch bị hạn chế theo thể lệ chương trình KM',
      1080: 'Giao dịch hoàn tiền bị từ chối. Giao dịch thanh toán ban đầu không được tìm thấy',
      1081: 'Giao dịch hoàn tiền bị từ chối. Giao dịch thanh toán ban đầu đã được hoàn',
      2001: 'Giao dịch thất bại do sai thông tin liên kết',
      2007: 'Giao dịch thất bại do liên kết thanh toán không tồn tại hoặc đã hết hạn',
      3001: 'Liên kết thanh toán bị từ chối vì người dùng chưa đăng ký dịch vụ',
      3002: 'Tài khoản chưa được kích hoạt',
      3003: 'Tài khoản đang bị khóa',
      4001: 'Giao dịch bị hạn chế theo thể lệ chương trình KM',
      4010: 'Giao dịch bị hạn chế do OTP chưa được gửi hoặc đã hết hạn',
      4011: 'Giao dịch bị từ chối vì OTP không hợp lệ',
      4100: 'Giao dịch thất bại do người dùng không xác nhận thanh toán',
      10: 'Hệ thống đang được bảo trì',
      99: 'Lỗi không xác định',
    };

    return messages[resultCode] || `Lỗi không xác định (code: ${resultCode})`;
  }
}

module.exports = new MoMoService();

File: hotel_mobile/lib/data/services/momo_service.dart
/**
 * MoMo Service - Xử lý thanh toán MoMo trong Flutter
 * 
 * Chức năng:
 * - Tạo payment request từ backend
 * - Mở trình duyệt/WebView để thanh toán
 * - Xử lý kết quả thanh toán
 * - Query trạng thái giao dịch
 */

import 'package:dio/dio.dart';
import '../../core/constants/app_constants.dart';
import '../../core/config/payment_config.dart';
import '../../core/services/backend_auth_service.dart';

/// Service xử lý MoMo
class MoMoService {
  final Dio _dio;
  final BackendAuthService _authService = BackendAuthService();

  MoMoService() : _dio = Dio(BaseOptions(
    baseUrl: AppConstants.baseUrl,
    connectTimeout: const Duration(seconds: 30),
    receiveTimeout: const Duration(seconds: 30),
    headers: {
      'Content-Type': 'application/json',
    },
  )) {
    // Add interceptor để thêm token vào header (optional - payment không require authentication)
    _dio.interceptors.add(InterceptorsWrapper(
      onRequest: (options, handler) async {
        final token = await _authService.getToken();
        if (token != null && token.isNotEmpty) {
          options.headers['Authorization'] = 'Bearer $token';
          print('✅ MoMo: Added token to header');
        } else {
          print('ℹ️ MoMo: Proceeding without authentication token');
        }
        return handler.next(options);
      },
    ));
  }

  /// Tạo payment URL từ MoMo (giống VNPay)
  /// 
  /// Parameters:
  /// - [bookingId]: ID của booking cần thanh toán
  /// - [amount]: Số tiền (VND)
  /// - [orderInfo]: Mô tả đơn hàng
  /// - [bookingData]: Thông tin booking để tạo sau khi thanh toán (optional)
  /// 
  /// Returns: Map với paymentUrl, qrCodeUrl, deeplink
  Future<Map<String, dynamic>> createPaymentUrl({
    required int bookingId,
    required double amount,
    required String orderInfo,
    Map<String, dynamic>? bookingData,
  }) async {
    try {
      print('📤 MoMo Service: Gửi request đến ${PaymentConfig.momoCreatePaymentUrlEndpoint}');
      print('📋 MoMo Service: bookingId=$bookingId, amount=${amount.toInt()}');
      print('📋 MoMo Environment: ${PaymentConfig.useMomoSandbox ? "Sandbox" : "Production"}');
      
      final response = await _dio.post(
        PaymentConfig.momoCreatePaymentUrlEndpoint,
        data: {
          'bookingId': bookingId,
          'amount': amount.toInt(), // MoMo yêu cầu số nguyên
          'orderInfo': orderInfo,
          if (bookingData != null) 'bookingData': bookingData,
        },
      );

      print('📥 MoMo Service: Nhận response - status=${response.statusCode}');
      print('📥 MoMo Service: response.data=${response.data}');

      if (response.data['success'] == true) {
        final data = response.data['data'];
        if (data == null) {
          throw Exception('Server trả về data rỗng');
        }
        if (data['paymentUrl'] == null || data['paymentUrl'].toString().isEmpty) {
          throw Exception('Server trả về payment URL rỗng');
        }
        print('✅ MoMo Service: Payment data nhận được thành công');
        return data as Map<String, dynamic>;
      } else {
        // Lấy error message từ server, ưu tiên message chi tiết
        String errorMsg = response.data['message'] ?? 'Không thể tạo payment URL';
        
        // Nếu có error code, thêm vào message
        if (response.data['error'] != null) {
          final errorCode = response.data['error'];
          if (errorCode == 'INVALID_RETURN_URL') {
            errorMsg = 'MoMo không chấp nhận localhost làm Return URL. Vui lòng set MOMO_RETURN_URL trong file .env với URL công khai.';
          }
        }
        
        print('❌ MoMo Service: Server trả về lỗi: $errorMsg');
        throw Exception(errorMsg);
      }
    } on DioException catch (e) {
      print('❌ MoMo Service: DioError - ${e.type}');
      print('❌ MoMo Service: Message: ${e.message}');
      print('❌ MoMo Service: Response: ${e.response?.data}');
      print('❌ MoMo Service: Status: ${e.response?.statusCode}');
      
      if (e.type == DioExceptionType.connectionTimeout || 
          e.type == DioExceptionType.receiveTimeout ||
          e.type == DioExceptionType.sendTimeout) {
        throw Exception('Kết nối quá thời gian. Vui lòng kiểm tra kết nối mạng và thử lại.');
      } else if (e.type == DioExceptionType.connectionError) {
        throw Exception('Không thể kết nối đến server. Vui lòng kiểm tra kết nối mạng.');
      } else if (e.response != null) {
        final errorMsg = e.response?.data['message'] ?? 
                        e.response?.data['error'] ?? 
                        'Lỗi từ server (${e.response?.statusCode})';
        throw Exception(errorMsg);
      }
      throw Exception('Không thể kết nối đến server. Vui lòng thử lại.');
    } catch (e) {
      print('❌ MoMo Service: Error không xác định: $e');
      if (e is Exception) {
        rethrow;
      }
      throw Exception('Lỗi không xác định: $e');
    }
  }

  /// Tạo payment request đến MoMo (legacy)
  /// 
  /// Parameters:
  /// - [bookingId]: ID của booking cần thanh toán
  /// - [amount]: Số tiền (VND)
  /// - [orderInfo]: Mô tả đơn hàng
  /// - [extraData]: Dữ liệu bổ sung (optional, base64 encoded)
  /// - [bookingData]: Thông tin booking để tạo sau khi thanh toán (optional)
  /// 
  /// Returns: Object chứa payUrl, deeplink, qrCodeUrl
  Future<Map<String, dynamic>> createPayment({
    required int bookingId,
    required double amount,
    required String orderInfo,
    String? extraData,
    Map<String, dynamic>? bookingData,
  }) async {
    try {
      final response = await _dio.post(
        '/api/v2/momo/create-payment', // Legacy endpoint
        data: {
          'bookingId': bookingId,
          'amount': amount.toInt(), // MoMo yêu cầu số nguyên
          'orderInfo': orderInfo,
          if (extraData != null) 'extraData': extraData,
          if (bookingData != null) 'bookingData': bookingData,
        },
      );

      if (response.data['success'] == true) {
        return response.data['data'];
      } else {
        throw Exception(response.data['message'] ?? 'Không thể tạo payment request');
      }
    } on DioException catch (e) {
      print('❌ DioError creating MoMo payment: ${e.message}');
      if (e.response != null) {
        print('❌ Response data: ${e.response?.data}');
        throw Exception(e.response?.data['message'] ?? 'Lỗi kết nối server');
      }
      throw Exception('Không thể kết nối đến server');
    } catch (e) {
      print('❌ Error creating MoMo payment: $e');
      throw Exception('Lỗi không xác định: $e');
    }
  }

  /// Query trạng thái giao dịch MoMo
  /// 
  /// Parameters:
  /// - [orderId]: Mã đơn hàng từ MoMo
  /// - [requestId]: Request ID từ MoMo
  /// 
  /// Returns: Map chứa thông tin giao dịch
  Future<Map<String, dynamic>> queryTransaction({
    required String orderId,
    required String requestId,
  }) async {
    try {
      final response = await _dio.post(
        PaymentConfig.momoQueryTransactionEndpoint,
        data: {
          'orderId': orderId,
          'requestId': requestId,
        },
      );

      if (response.data['success'] == true) {
        return response.data['data'];
      } else {
        throw Exception(response.data['message'] ?? 'Không thể truy vấn giao dịch');
      }
    } on DioException catch (e) {
      print('❌ DioError querying MoMo transaction: ${e.message}');
      if (e.response != null) {
        throw Exception(e.response?.data['message'] ?? 'Lỗi kết nối server');
      }
      throw Exception('Không thể kết nối đến server');
    } catch (e) {
      print('❌ Error querying MoMo transaction: $e');
      rethrow;
    }
  }
}

